lib/authenticate_methods.rb

```diff
@@ -0,0 +1,111 @@
+module AuthenticateMethods
+
+  def authenticate_user!
+    use_signed_params if signed_params.present?
+
+    return if signed_in?
+
+    # Drop signed params from where we will go back after log in so we don't
+    # try to use them again.
+    store_url(url: request_url_without_signed_params)
+
+    if signup_state && signup_state.trusted_student?
+      redirect_to main_app.signup_path
+    else
+      redirect_to(
+        main_app.login_path(
+          params.slice(:client_id, :signup_at, :go, :no_signup)
+        )
+      )
+    end
+  end
+
+  def authenticate_admin!
+    return if current_user.is_administrator?
+
+    store_url
+    redirect_to main_app.login_path(params.slice(:client_id))
+  end
+
+  # Doorkeeper controllers define authenticate_admin!, so we need another name
+  alias_method :admin_authentication!, :authenticate_admin!
+
+  protected
+
+  def use_signed_params
+    incoming_user = nil
+    found_incoming_user_by = nil
+
+    # First try to to find an existing user by an external UUID
+
+    if external_user_uuid.present?
+      incoming_user = UserExternalUuid.find_by_uuid(external_user_uuid).try(:user)
+
+      if incoming_user.present?
+        found_incoming_user_by = "external uuid"
+      end
+    end
+
+    # If that didn't pan out, try to find an existing user with a verified email
+    # matching the secure params email
+
+    if incoming_user.nil? && external_email.present?
+      incoming_user = LookupUsers.by_verified_email(external_email).first
+
+      if incoming_user.present?
+        found_incoming_user_by = "external email"
+
+        if external_user_uuid.present?
+          # Go ahead and track the external UUID so can use in future
+          incoming_user.external_uuids.create!(uuid: external_user_uuid)
+        end
+      end
+    end
+
+    if incoming_user.present?
+      if signed_in? && incoming_user != current_user
+        # Sign out the current user; signing in the new user will effectively sign out
+        # the current user, but we choose to sign out explicitly so we can record
+        # information in the current user's security log
+
+        sign_out!(security_log_data: {type: 'different external user'})
+      end
```

Comment:
I think there's a small loophole here.

* If the user is signed out, we search for a matching validated email and use that if found.
* However if they're signed in, we force them to go through the sign in/up flow, regardless of email matching.

I think we also need to check the validated emails of a signed-in user and link/redirect back in that case also.

Reply:
discussed with @nathanstitt and we think this is ok.